test/test_decomp.py

```diff
@@ -0,0 +1,481 @@
+    ):
+        return FLOAT64_DTYPE
+    else:
+        return x
+
+
+def normalize_op_input_output(f, sample, requires_grad=True):
+    args = tuple([sample.input] + list(sample.args))
+    return normalize_op_input_output2(
+        f,
+        args,
+        sample.kwargs,
+        sample.output_process_fn_grad,
+        requires_grad=requires_grad,
+    )
+
+
+CROSS_REF_EXCLUDE_SET = {
+    (
+        "cpu",
+        torch.bfloat16,
+        "nn.functional.layer_norm",
+    ),  # "batch_norm" not implemented for 'BFloat16'
+    ("cpu", torch.bfloat16, "addmm"),  # decomposition loses precision
+    ("cpu", torch.bfloat16, "softmax"),  # needs relaxed prec
+    ("cpu", torch.bfloat16, "log_softmax"),  # needs relaxed prec
+    # decomp has problem even with opmath
+    ("cuda", torch.bfloat16, "nn.functional.layer_norm"),
+    ("cuda", torch.float16, "nn.functional.layer_norm"),
+    ("cuda", torch.float16, "nn.functional.dropout"),
+    ("cuda", torch.bfloat16, "nn.functional.dropout"),
+    # decomp doesn't return correct dtype
+    ("cuda", torch.float64, "nn.functional.instance_norm"),
+    ("cuda", torch.float32, "nn.functional.instance_norm"),
+    ("cuda", torch.float64, "nn.functional.dropout"),
+    ("cuda", torch.float32, "nn.functional.dropout"),
+    ("cuda", torch.float64, "nn.functional.batch_norm"),
+    ("cuda", torch.float32, "nn.functional.batch_norm"),
+    # complex is not handled
+    (None, torch.complex64, "var"),
+    (None, torch.complex128, "var"),
+    (None, torch.complex64, "nn.functional.tanhshrink"),
+    (None, torch.complex128, "nn.functional.tanhshrink"),
+    (None, torch.complex32, "sigmoid"),
+    (None, torch.complex64, "sigmoid"),
+    (None, torch.complex128, "sigmoid"),
+    (None, torch.complex64, "tanh"),
+    (None, torch.complex128, "tanh"),
+}
+
+all_decomposed = set()
+
+# Helpful snippet for testing coverage
+"""
+import atexit
+def check_coverage():
+    print("missing coverage:")
+    print("\n".join(map(str, decomposition_table.keys() - all_decomposed)))
+atexit.register(check_coverage)
+"""
+
+
+class TestDecomp(TestCase):
+    longMessage = True
+
+    # NB: This actually overlaps with test_comprehensive, but it only
+    # runs on things that are definitely decomposed so it's a lot faster
+    # to run
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(_decomp_test_ops)
+    def test_quick(self, device, dtype, op):
+        self.do_cross_ref(device, dtype, op, run_all=False)
+
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_comprehensive(self, device, dtype, op):
+        self.do_cross_ref(device, dtype, op, run_all=True)
+
+    def do_cross_ref(self, device, dtype, op, *, run_all):
+        if (torch.device(device).type, dtype, op.name) in CROSS_REF_EXCLUDE_SET or (
```

Comment:
fyi maybe these skips could be modeled as skips on the OpInfos themselves for consistency?

Reply:
I guess the issue with that is you'd have to skip multiple tests, which is kinda a pain?

Reply:
I thought about it... it just seemed a lot harder to handle the skips if they were smeared out through the OpInfo database, rather than centralized on the test itself. I guess this is a general point on OpInfo skips: I think they should be attached to the relevant tests, rather than the OpInfo entries themselves.